### PR TITLE
Sparse Gradient Clipping

### PR DESCRIPTION
**Pull request by vidurj**
Currently gradient clipping fails with `clamp is not implemented for type torch.sparse.FloatTensor` if any of the parameters have sparse gradients. This PR ignores them for gradient clipping purposes.

### CODE REVIEW

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -348,7 +348,10 @@ def _enable_gradient_clipping(self) -> None:
         if self._grad_clipping is not None:
             # Pylint is unable to tell that we're in the case that _grad_clipping is not None...
             # pylint: disable=invalid-unary-operand-type
-            clip_function = lambda grad: grad.clamp(-self._grad_clipping, self._grad_clipping)
+            def clip_function(grad):
+                # Sparse tensors do not have a clamp method.
```
Comment (matt-gardner) [inline allennlp/training/trainer.py]: Is there no way to get the same functionality with sparse tensors?  I feel like this should at least emit a warning if it can't do what you want, instead of silently failing.

Comment (matt-peters) [reply allennlp/training/trainer.py]: You should be able to clamp the values directly and create a new sparse tensor to return.

Comment (vidurj) [reply allennlp/training/trainer.py]: True, but do I need to be careful about coalescing the sparse tensor before clamping the values?

Comment (matt-peters) [reply allennlp/training/trainer.py]: I don't think so.  See `sparse_clip_norm` for an example that clips gradient norms of sparse gradients.

Comment (vidurj) [reply allennlp/training/trainer.py]: Thanks!

**Comment by schmmd** [comment]
@vidur-joshi can this be merged or are you waiting on review from @matt-gardner?

**Comment by vidurj** [comment]
Oops, I forgot it hadn't been merged. I'll merge it when the tests pass.
